accepted/2021/architecture-targeting.md

```diff
@@ -0,0 +1,498 @@
+
+The proposal aims to resolve the outstanding problems with architecture targeting, not solely the aspects required to satisfy the x64 emulation scenario. However, the proposal is prioritized in terms of the x64 emulation scenario, as described near the end of the document.
+
+## Context
+
+Today, architecture targeting shows up in the following SDK experiences and concepts:
+
+* Deployment models: framework-dependent vs self-contained.
+* RID targeting: RID-neutral vs RID-specific.
+* Various CLI verbs and their capabilities.
+
+Across these topic areas, we find that the .NET SDK lacks a coherent model for architecture targeting. The following statements provide a brief description of the SDK UX today. Feel free to skip this section. It is primarily here to demonstrate the state of the current model.
+
+### Good
+
+* The SDK produces framework-dependent apps by default across all verbs. These apps rely on a separately installed runtime on the machine of a compatible version.
+
+### Bad
+
+* The SDK produces portable apps by default across all verbs. These apps  can run on multiple operating systems and architectures.
+* At the same time, the SDK generates a native "apphost" executable for the app that matches the RID of the SDK used. This isn't symmetric with the app being portable, as the apphost will only work in one of the environments that the app will. There is no experience to reason about that (which single environment does the apphost support? which environments does the underlying app support?)
+* The presence of the apphost for a portable app might be useful for development, but it is confusing in a deployed environment if the development and production RIDs do not match. If they do match, then the app isn't as optimized as it could be.
+* Portable apps will be bigger if the app depends on packages with architecture-specific dependencies, since the app will (likely) contain the native assets for multiple architectures (that's one of three key aspects of what makes it portable). The only reason to build portable apps is to accommodate multi-architecture packages.
+* Portable apps and RID-specific builds [do not place native assets in the same location](https://github.com/dotnet/sdk/issues/11754). That should be configurable and is a [confusing behavior](https://github.com/dotnet/sdk/issues/3903).
+* It is possible to make an app portable for one operating system, such as for Windows x86, x64, and Arm64, but that isn't enabled.
+* Portable apps -- for the apphost -- don't support the universal binaries feature of Windows or macOS, which reduces their utility, particularly for client apps.
+* Some users want to target a specific architecture and/or operating system. The SDK produces a self-contained app by default when you do this. Many users (likely most) do not want that.
```

Comment:
Totally agree. I think most people don't (A) realize that this model even exists, (B) don't realize why they would use it, and (C) don't understand the syntax. My take is that RID specific fx-dependent apps are the best default for the product. That's much of the thesis of this proposal.

Reply:
I wonder if we can get some data around which deployments are used more often.

I would hope that the more used scenarios would be easier for the user - less settings to pass to the command line / set in the .csproj.

Reply:
I think getting deployments for real world desktop apps is also important.

What you want for a simple command line tool and what you want for a game or a more advanced app might be very different. I'm not sure how well this model fits for Paint.NET for example where, from what I've seen from Rick brew in the C# Community Discord, it seems to be a bit of a different setup and deployment model.

Reply:
It would be very hard to reason about usage patterns/data and I don't think it was lead us to the right decision, in part due to what @tannergooding raises.

The current system doesn't align well with user needs. I don't think most users want portable by default and the focus on `-r` is not good UX. Instead, users can focus on `--self-contained` and specialize it with a architecture/OS info as an option.  I don't see data pointing us in that direction given how challenging the current UX is.

Reply:
Also sometimes you want self contained without only AnyCPU parts of the BCL and not the complete runtime that is cpu specific too (and so you would require them to need to have the cpu specific runtime bits either installed or in separate folders within an ``runtime`` subfolder within the self contained application where it would *somehow* select the right bit of runtime to load the thing).

Reply:
> I don't see data pointing us in that direction given how challenging the current UX is.

Data isn't going to point us to the right UX. But data is going to point us to "what do users want to do more often?", which will help us create the right UX.

Making users specify 3 arguments instead of 1 in a hardly used scenario isn't that big of a deal. But making users specify 3 arguments instead of 1 in an often used scenario is a big deal IMO.

`dotnet publish -r win-x64`

vs.

`dotnet publish --self-contained --os win -a x64`

Which would you rather type?

Reply:
Sure, but that's cherry-picking the proposal.

The new proposal does the following, across .NET 6 and 7:

- Improves the default (portable fx-dependent to rid-specific fx-dependent).
- Improves self-contained build UX when the target RID and the SDK RID match.
- Regresses self-contained build UX when the target RID and SDK RID do not match.

Do you agree with that assessment?

Reply:
I agree that the proposal does that. And I definitely agree I want the first two.

But the proposal also does:

* When the target RID and SDK RID do not match, prioritize fx-dependent apps over self-contained 

That is the part of the proposal that I don't want for myself as a user. And so that's where my ask for data/evidence is coming from.

In my experience, when I publish for a RID that isn't the SDK RID, every time I want self-contained. I honestly can't think of a time when I've wanted a fx-dependent app. So I want to verify that most users do want that behavior.

Reply:
In practice, I'm not typing either. I'm coding this up in a CI script so I don't need to remember what to type everytime and relying on F5 debugging for local testing.
If I need to test the "shipping product" I'll submit a PR to my repo and have CI spit out some artifacts for me to try out (or I can simulate that experience locally with `.\scripts\cibuild.cmd` for my own repos and that takes care of all the options)

I'm not a big enterprise and as an OSS project likely don't have the ability to cover all the CDN and deployment costs for all the self-contained binaries. So for me, I'd ideally be able to produce small portable deployments that the installer can then finish doing things like AOT or R2R on at install time so that startup is fast and its taking advantage of all the cool features .NET does have available. This helps with any costs I need to pay, fits well within the limitations of most free/low tier services (e.g. NuGet package size or GitHub artifact size limitations), and avoids needing to maintain or explicitly support deployments for every OS (windows, linux, macos, android, ios, etc), every distro (win7, win10, win11, ubuntu18.04, ubuntu20.04, rhel, suse, freebsd, etc), and every architecture (x86, x64, Arm32, Arm64, etc) or needing to consider what the "average" hardware a target user has for cases where myself or a dependency use hardware acceleration (SSE4.1, AVX, AVX2 on x86/x64 or AdvSIMD and friends on A64, or WebSIMD for WASM in the future, etc).

Reply:
-- For reference, the above is also not be speaking abstractly. I maintain several OSS libraries and tools independently of work and these are the kinds of problems I actively run into already.

Reply:
> I'd ideally be able to produce small portable deployments that the I can then finish doing things like AOT or R2R on at install time or first launch so that startup is fast

This is separate from the overall conversation, but we don't intend to build those experiences. We're focused on dev-time crossgen and awesome JIT features w/nothing in-between. Today, crossgen has a couple characteristics that we really like:

- The build produces final optimized artifacts, per OS/arch.
- Those artifacts can be signed.

The NGEN experience had neither of those characteristics and had a ton of UX problems. We're not going to recreate those challenges.

> I honestly can't think of a time when I've wanted a fx-dependent app. So I want to verify that most users do want that behavior.

The thinking behind this change is super simple (as I articulate in the doc):

- fx dependent is the default build type, since .NET Core 1.0.
- rid-specific is relevant to fx-dependent apps, using an implicit RID, and this will be the default in .NET 7.
- specifying an explicit rid shouldn't change the deployment type.

I respect and sympathize with the point you are raising. It's legit. My view is coming from a very different direction, as I've articulated. I am a fan of self-contained. Judging by our container image pulls (10s of millions/month), self-contained is not the primary developer scenario.

I think the new model is generally better. It's not better for all cases but that's typical of change.

Also, for the x64 emulation scenario, rid-specific targeting absolutely benefits from self-contained not being the default.

Reply:
> This is separate from the overall conversation, but we don't intend to build those experiences.

I think that's unfortunate and is overlooking the independent developer scenario. In particular certain things like signing simply aren't a consideration for many independent developers. Authenticode certs tend to start at $500-600/yr and that's out of reach for a lot of projects (assuming they aren't in the .NET Foundation or similar). Having the ability to easily perform steps at any stage, including post-install, is general goodness IMO and allows developers to setup their CI flow however works best for them, even if they are signing, etc.

Reply:
Sure. I'd rather us focus on dynamic PGO than a post-install NGEN replacement. BTW: I don't think that NGEN was very popular either, particularly among the audience you are describing.

Reply:
I think that might have been for other reasons and that dynamic PGO isn't really a viable alternative for handling fast startup, its more about tuning the running application to achieve higher end throughput.

I think its worth considering what would be required to make our core build tools, including `crossgen`, the `linker`/`trimmer`, and others usable either standalone or as part of the SDK. I imagine its not much and that its mostly just packaging them as `dotnet tools`. This would cover the normal "publish" scenario, running them as post-build or independent CI steps, or even integration into custom build pipelines (which frameworks like Unity have).